x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,340 @@
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.Mapper;
+import org.elasticsearch.index.mapper.MapperParsingException;
+import org.elasticsearch.index.mapper.ParseContext;
+import org.elasticsearch.index.mapper.TypeParsers;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.query.QueryShardException;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** A {@link FieldMapper} for a field containing aggregate metrics such as min/max/value_count etc. */
+public class AggregateMetricFieldMapper extends FieldMapper {
+
+    public static final String CONTENT_TYPE = "aggregate_metric";
+
+    /**
+     * Mapping field names
+     */
+    public static class Names {
+        public static final String IGNORE_MALFORMED = "ignore_malformed";
+        public static final ParseField METRICS = new ParseField("metrics");
+    }
+
+    enum Metrics {
+        min, max, sum, value_count;
+    }
+
+    public static class Defaults {
+        public static final Explicit<Boolean> IGNORE_MALFORMED = new Explicit<>(false, false);
+        public static final Explicit<Set<String>> METRICS = new Explicit<>(Collections.emptySet(), false);
+        public static final AggregateMetricFieldType FIELD_TYPE = new AggregateMetricFieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setHasDocValues(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class Builder extends FieldMapper.Builder<AggregateMetricFieldMapper.Builder, AggregateMetricFieldMapper> {
+
+        private Boolean ignoreMalformed;
+
+        private Set<String> metrics;
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        public AggregateMetricFieldMapper.Builder ignoreMalformed(boolean ignoreMalformed) {
+            this.ignoreMalformed = ignoreMalformed;
+            return builder;
+        }
+
+        protected Explicit<Boolean> ignoreMalformed(BuilderContext context) {
+            if (ignoreMalformed != null) {
+                return new Explicit<>(ignoreMalformed, true);
+            }
+            if (context.indexSettings() != null) {
+                return new Explicit<>(IGNORE_MALFORMED_SETTING.get(context.indexSettings()), false);
+            }
+            return AggregateMetricFieldMapper.Defaults.IGNORE_MALFORMED;
+        }
+
+        public AggregateMetricFieldMapper.Builder metrics(Set<String> metrics) {
```

Comment:
Should we use an `Enum` here to ensure that we restrict to metrics that we can handle (e.g.: `max`, `min`, `avg`) ?

Reply:
Fixed